### PR TITLE
Add proxy support

### PR DESCRIPTION
**Pull request by T0jan**
A small change concerning parameter handling. Consists of two parts:

* The ECCN information from Digikey will be processed as parameter like it is for Element14
* load_category_parameters is restructured to support parameter mappings for categories further down the tree as well as multiple mappings in a tree. 

also contains the commits of #146 because I am too stupid for correct branching :sweat_smile:

### CODE REVIEW

**Comment by eeintech** [comment]
Hello @T0jan 

Do you mind pulling main branch into yours now that I've just released `1.0.0`?

What about if we closed #146 and add the description of proxy changes here instead? That way I'll check out this PR only :smiley:

**Comment by T0jan** [comment]
@T0jan congrats on the release :D

I just synced my branch so no problems on this side.

The only reason I splitted this into a different merge request is, if the change to the parameter mapping conflicts with anything you meant to do with it or if you don't like my solution there, then we could include the proxy stuff independently. So do as you please :D

**Comment by eeintech** [comment]
@T0jan Thanks, I'll have a look this week :+1:

**Comment by eeintech** [comment]
> this is to be expected. basically the GUI now sends proxy request to your localhost server which this one can't interpret as it is not configured as a proxy server. if you want to test with a proxy server you can use any from a free proxy server, e.g. one from [here](http://free-proxy.cz/en/)

Alright so I used one of the servers, put a random token (else it asks for username) and I'm getting this:
``` shell
Server connection error: <class 'requests.exceptions.SSLError'>
[TREE]	Successfully connected to InvenTree server (ENV=DEVELOPMENT)
```

So not sure if that actually works...
Do you think we can implement a proxy test in `run_tests.py` using one of those free proxy servers?

> we could add a special error message based on the requests module error case `Cannot connect to proxy` not sure if it would trigger in this case tho.

I guess for now the general message is clear, something is wrong with the InvenTree settings.
Let's focus on this later.

> Yes, this is expected. if you want to connect to an HTTP-server it will use the HTTP-proxy and for a HTTPS-server the HTTPS-proxy. thinking about it probably would make sense here to give the user only one proxy input field here and letting the program decide (based on the server URL) if it is HTTP or HTTPS. do you think this would make things clearer for the user?

One field would definitely be clearer than two and lead to less confusion I believe. If you can use a generic name like `Proxy URL` that could do it for me.
One question I have for you, if you enter a Proxy URL, do you need to enter anything else?

**Comment by eeintech** [comment]
Hello again @T0jan 

Regarding those two items:
> * The ECCN information from Digikey will be processed as parameter like it is for Element14
> * load_category_parameters is restructured to support parameter mappings for categories further down the tree as well as multiple mappings in a tree.

I understand you'd like the `export_control_class_number` to be part of the Digi-Key search data, however I think your implementation is quite crude and not cleanly expandable to other data fields, and it's all my fault not to have made it more flexible so I would like to propose to fix it :smiley: 
Looking into it, I think adding a separate dataset of `extra_fields` that can be controlled via each supplier `_config.yml` configuration file could work. In the config file, one would just have to add `export_control_class_number` to the list under the "EXTRA_FIELDS" key (inside the config file) to get the result as part of the search result, and all those extra fields would be treated as parameters.
What do you think?

Regarding the `load_category_parameters`, it seems fine to me and looks like a sensible change, thanks for adding this!

**Comment by T0jan** [comment]
hi @eeintech 
> Alright so I used one of the servers, put a random token (else it asks for username) and I'm getting this:

did you tried with different proxy servers and to which inventree server do you wanted to connect? The servers in the list often don"t work so you have to try a bit to find a working one. I managed to get a connection to the inventree-demo server with:

```
Server Address: https://demo.inventree.org
Username: reader
Password: readonly
HTTPs Proxy: https://167.172.173.210:36917
```

as well as:
```
HTTPs Proxy: socks5h://104.237.134.201:59166
```

> Do you think we can implement a proxy test in run_tests.py using one of those free proxy servers?

I am not sure how reliable these proxy servers are. I usually use them only for immediate testing and have sometimes to search a while to find a working one. So we could often get errors in run_tests.py only because the proxy is not working.

> One field would definitely be clearer than two and lead to less confusion I believe. If you can use a generic name like Proxy URL that could do it for me.

Maybe it would be even a better idea to hide this field behind an additional switch, similar to the input fields for the IPN. :thinking: 

> One question I have for you, if you enter a Proxy URL, do you need to enter anything else?

No you just have to construct the URL correctly like in my two examples above: **protocol**://**address**:**port**

> Looking into it, I think adding a separate dataset of extra_fields that can be controlled via each supplier _config.yml configuration file could work. In the config file, one would just have to add export_control_class_number to the list under the "EXTRA_FIELDS" key (inside the config file) to get the result as part of the search result, and all those extra fields would be treated as parameters.
What do you think?

Sounds like a lot cleaner approach then my try to press it into the current structure :smile: 

The only problem I see here is that you then get two config files (the supplier_parameters.yaml and the *supplier*_config.yaml) the user needs to modify to get the data actually into inventree. but maybe the *supplier*_config.yaml can hold the most common extra fields a user would need by default so only special cases would have to look into this file?

**Comment by T0jan** [comment]
at some point I will learn to check linting before I push...

@eeintech I just added the discussed hiding of the proxy setting. for the extra_fields do you have already an idea how you want to do it? then I would revert my commits related to the ECCN field and you can work cleanly on it in your repo.

**Comment by eeintech** [comment]
> @eeintech I just added the discussed hiding of the proxy setting. for the extra_fields do you have already an idea how you want to do it? then I would revert my commits related to the ECCN field and you can work cleanly on it in your repo.

I sure can do that, if you don't mind reverting that change.

About the proxy implementation, the settings look good:

image

Without proxy it works great.
However, I cannot find a proxy which works in this list: http://free-proxy.cz/en/
Did you find a good working one?

Also, how do you test connecting with a token only? Where do you get the token from?

File: kintree/gui/views/settings.py [hunk 1]
```diff
@@ -732,13 +745,24 @@ class InvenTreeSettingsView(SettingsView):
     ]
 
     def save(self, file=None, dialog=True):
+        address = SETTINGS[self.title]['Server Address'][1].value
+        proxy = SETTINGS[self.title]['Proxy'][1].value
+        enable_proxy = SETTINGS[self.title]['Enable Proxy Support'][1].value
+        if not enable_proxy:
+            proxies = {}
```
Comment (eeintech) [inline kintree/gui/views/settings.py:752]: Could you replace this line with:
``` python
proxies = None
```

else it saves it as `PROXIES: {}` in config file which I don't find very clean

**Review by eeintech — CHANGES_REQUESTED** [review]

**Comment by eeintech** [comment]
Hello @T0jan 

Couple comments in your PR, could you please review?

Other than that I'm happy to merge :smiley:

**Comment by eeintech** [comment]
Could you also rebase on main? it created a lot of duplicate commits :confused:

**Comment by eeintech** [comment]
> @eeintech 2nd rebase went better, should have a clean commit history now.
> 
> I modified the README and cleaned up the Proxy settings and definitions in case no proxy is used as requested.

Thanks, I'll take a look!

> On the token topic: I was thinking about adding a third button next to save to replace the user credentials with the token automatically as all requests in the background are anyway done with the token method. any opinions on this?

So a "Get Token" button? But this will overwrite the password/token field and then it won't be valid next time Ki-nTree is opened... am I understand this correctly? Would this button useful for proxy method only, because I don't see why using it when you have your username/password filed in?

**Comment by eeintech** [comment]
Everything looks good to me and I'm happy to merge as-is unless you really think adding the "Get Token" button is necessary to simplify the proxy workflow.

**Review by eeintech — APPROVED** [review]

**Comment by eeintech** [comment]
@T0jan Thanks for the feedback. Let's go with what you proposed so far, we can discuss the extra button for another release, if really necessary.

Thanks a lot for your contribution :+1:

**Pull request merged**